### PR TITLE
`General`: Include more information in course archive

### PR DESCRIPTION
**Pull request by tobias-lippert**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
### Motivation and Context

Course archives do not contain exercise details or the problem statement for all exercise types. This PR adds this information to course archives.

### Description
Refactored the `SubmissionExportService`. Fixed the exports of File Upload submissions. Introduced a new ExerciseWithSubmissionsExportService that exports a JSON file with information about the exercise, a markdown file with the problem statement and all submissions or student repos for an exercise.  For programming exercise, we also export template, solution, test, and aux repositories.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to Course Administration
3. Select a course that has an end date in the past or set the end date of a course to a value in the past.
4. Click on archive
5. Once finished, download the archive and verify its content.
6. Every exercise should have its own directory with the problem statement as markdown file, the exercise details json file and all submisisons.
7. Also test the export of submissions (for text, modeling, and file upload) available on the scores page of an exercise (empty submissions are always excluded)
8. Also test that the export of a single programming exercise (using the download exercise button) still works.
9. Run a plagiarism check for a programming exercise and check it still works (because it's affected by the changes in FileService)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [CourseExamExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5420/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/CourseExamExportService.html) | 90% | ✅|
| [ExerciseWithSubmissionsExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5420/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/ExerciseWithSubmissionsExportService.html) | 91% | ✅|
| [FileUploadExerciseWithSubmissionsExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5420/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/FileUploadExerciseWithSubmissionsExportService.html) | 100% | ✅ |
| [ModelingExerciseWithSubmissionsExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5420/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/ModelingExerciseWithSubmissionsExportService.html) | 100% | ✅ |
| [TextExerciseWithSubmissionsExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5420/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/TextExerciseWithSubmissionsExportService.html) | 100% | ✅|
| [ProgrammingExerciseExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5420/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.export/ProgrammingExerciseExportService.html) | 85% | ✅|

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/CourseExamExportService.java [hunk 1]
```diff
@@ -406,21 +412,21 @@ else if (exercise instanceof QuizExercise) {
...
-                    exportedExercises.add(newExportedSubmissionsFilePath);
-
-                    // Delete the directory where the zip was located before it was moved
-                    FileUtils.deleteDirectory(Path.of(exportedSubmissionsFile.getParent()).toFile());
-                }
-                catch (IOException e) {
-                    logMessageAndAppendToList("Failed to move file " + exportedSubmissionsFile.toPath() + " to " + outputDir + ".", exportErrors, e);
-                }
-            }
+            // if (exportedSubmissionsFileOrEmpty.isPresent()) {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/CourseExamExportService.java:415]: Can you remove this or why is this here?

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 2]
```diff
@@ -433,8 +433,8 @@ public String actualPathForPublicPath(String publicPath) {
      *
      * @param actualPathString the path to the file in the local filesystem
      * @param entityId         the id of the entity associated with the file
-     * @throws FilePathParsingException if the path is unknown
      * @return the public file url that can be used by users to access the file from outside
+     * @throws FilePathParsingException if the path is unknown
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:437]: Why doesn't the signature contain this exception?

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 3]
```diff
@@ -1162,6 +1162,7 @@ public Path writeStringToFile(String stringToWrite, Path path) {
             outStream.write(stringToWrite);
         }
         catch (IOException e) {
+            e.printStackTrace();
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:1165]: Teamscale

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 4]
```diff
@@ -590,6 +557,14 @@ private Path createZipForRepository(VcsRepositoryUrl repositoryUrl, String zipFi
         return gitService.zipFiles(localRepoPath, zipFilename, outputDir.toString(), contentFilter);
     }
 
+    private Path exportPlainRepository(VcsRepositoryUrl repositoryUrl, Path outputDir) throws GitAPIException, GitException, UncheckedIOException {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:560]: unused

File: src/test/java/de/tum/in/www1/artemis/course/CourseTestService.java
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/course/CourseTestService.java:1]: see teamscale

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 2 above)
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:437]: removed because unrelated to my changes.

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by JohannesStoehr — DISMISSED** [review]
Approve code

**Comment by Xinyannnnnn** [comment]
Tested on TS3 and it works as expected.

File: src/main/java/de/tum/in/www1/artemis/service/export/ProgrammingExerciseExportService.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/export/ProgrammingExerciseExportService.java:1]: This file has several Codacy warnings concerning missing JavaDoc and Missing Doc tags

File: src/main/java/de/tum/in/www1/artemis/service/export/SubmissionExportService.java [hunk 5]
```diff
@@ -95,13 +100,13 @@ public Optional<File> exportStudentSubmissions(Long exerciseId, SubmissionExport
      * @return a reference to the zipped file
      * @throws IOException if an error occurred while zipping
      */
-    public Optional<File> exportStudentSubmissions(Long exerciseId, SubmissionExportOptionsDTO submissionExportOptions, Path outputDir, List<String> exportErrors,
-            List<ArchivalReportEntry> reportData) throws IOException {
+    public List<Path> exportStudentSubmissions(Long exerciseId, SubmissionExportOptionsDTO submissionExportOptions, boolean zipSubmissions, Path outputDir,
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/export/SubmissionExportService.java:103]: JavaDoc

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java [hunk 6]
```diff
@@ -267,7 +268,7 @@ public ResponseEntity<Resource> exportInstructorExercise(@PathVariable long exer
         long start = System.nanoTime();
         Path path;
         try {
-            path = programmingExerciseExportService.exportProgrammingExerciseInstructorMaterial(programmingExercise, Collections.synchronizedList(new ArrayList<>()));
+            path = programmingExerciseExportService.exportProgrammingExerciseForDownload(programmingExercise, Collections.synchronizedList(new ArrayList<>()));
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java:271]: Why can't this be a `concurrentList` instead of  a `synchronizedList`?

**Review by MaximilianAnzinger — COMMENTED** [review]
Left some comments

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java (hunk 6 above)
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java:271]: tbh, I didn't introduce this but @krusche did in #7165.

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by Xinyannnnnn — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java (hunk 6 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java:271]: I found this online to be the best solution for concurrent threads. Tbh, I do not really know the differences between synchronizedList and synchronizedList

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java:271]: as long as it works, and there are no big disadvantages for synchronizedList, I would leave it

Comment (MaximilianAnzinger) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseExportImportResource.java:271]: concurrentList allows multiple threads to perform R/W-OPs concurrently. For synchronizedList this is limited to only one thread at a time.
Leaving it like this is fine with me 👍

**Review by MaximilianAnzinger — DISMISSED** [review]
Code

**Review by MaximilianAnzinger — APPROVED** [review]
Code

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
Manual: Tested on TS1. The only thing that didn't work was the archive of solution/template/test repositories. Everything else works like a charm.

**Review by Strohgelaender — APPROVED** [review]
code lgtm.

Tested on ts1, the issue mentioned above is fixed and everything else worked as expected.

**Review by MaximilianAnzinger — DISMISSED** [review]
Code and manual on TS1

Thx for the quick fix 👍

**Review by laurenzfb — APPROVED** [review]
Tested on ts1, worked as described

**Review by Strohgelaender — APPROVED** [review]
the archive contains the correct information and the plagiarism checks also still work.

**Review by jakubriegel — APPROVED** [review]
Tested on ts4. Everything including export of the course, single prog. exercise and plagiarism detection works fine 👍🏻

**Review by DominikRemo — APPROVED** [review]
Tested on ts2. Works as expected

**Review by MaximilianAnzinger — APPROVED** [review]
